src/architecture/merkledb.md

```diff
@@ -0,0 +1,348 @@
+possible to plug in other engines.
+
+To use a particular database, a minimal [`Database`][database]
+interface should be implemented for it:
+
+- Get a value by a [column family][col-family] name and a key
+- Put a new value at the specified column family / key (insert or update
+  the saved one)
+- Delete a key-value pair by a column family name / key.
+
+All the objects functionality is reduced to these atomic call types.
+Values of items of different objects are stored in a single column family in
+the low-level storage. Their keys are mapped to the low-level storage keys in a
+deterministic manner using [object identifiers](#object-identifiers).
+
+### Object Identifiers
+
+On user level every object is uniquely identified by an address. An object
+address consists of 2 parts:
+
+- **String name** that may contain uppercase and lowercase Latin letters,
+  digits, underscores `_`, hyphens `-` and periods `.`. By convention, object
+  names in services should start with [the service name][service-name] and a
+  period. For example, the only object in the Cryptocurrency Tutorial is named
+  `cryptocurrency.wallets`, where `cryptocurrency` is the service name, and
+  `wallets` is the own name of the object
+- **Optional prefix** presented as a sequence of bytes (`Vec<u8>` in Rust
+  terms). Prefixes allow to group items inside objects. They allow to obtain
+  a particular subset of items marked by the prefix from the object.
+
+The object address maps the object to the corresponding unique identifier. To
+obtain object identifiers [object pool][indexes-pool] is used. The pool
+stores identifiers of all available objects. The pool assigns identifiers to
+the objects incrementally as soon as new objects appear.
+
+The pool is stored in a separate column family. Key in this column family is
+the object address presented in bytes, value is the
+[`IndexMetadata`][index-metadata] structure, which stores the object
+identifier and some object metadata.
+
+The object metadata can store various information about the internal state of
+the object. For example, for [`ListIndex`], the metadata stores the length of
+the list.
+
+An object key in RocksDB consists of two parts:
+
+- an **object identifier** obtained from the object pool
+- an **index of an item** inside the object.
+
+Both parts of the key are encoded as big-endian.
+
+!!! note "Example"
+    Suppose we have a list with the address `(" exchange.crypto "," BTC ")` in
+    which we put one value, for example `7865`. The pool assigns identifier
+    `3` to this list when we create it. In the database the identifier looks
+    like this - `0x0000000000000003`.
+
+    Since `7865` is a single value in the created list, its index in the
+    database looks like this - `0x0000000000000000`.
+
+    Thus, the whole key in the database will look as follows (in HEX) -
+    `0x00000000000000030000000000000000 : 0x302E3030303139`.
```

Comment:
> `0x302E3030303139`

? This is ASCII / UTF-8 encoding of a string `0.00019`. How does it correspond to 7865? As an example, 7865 as a 64-bit signed (or unsigned) little-endian integer is encoded as `0x_bd1e_0000_0000_0000`.

Reply:
corrected